File: LeetCode/Medium/0049-group-anagrams/0049-group-anagrams.go
func groupAnagrams(strs []string) [][]string {
    var prime = []int64{2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,79,83,89,97,101}
    ret := make([][]string, 0)
    group := make(map[int64][]string)
    
    for i := 0; i < len(strs); i++ {
        var hash int64 = 1
        
        for j := 0; j < len(strs[i]); j++{
            hash = hash * prime[strs[i][j] - 'a'] % 10000000007
        }
        
        if _, ok := group[hash]; !ok { group[hash] = make([]string, 0) }
        group[hash] = append(group[hash], strs[i])
    }
    
    for _, v := range group { ret = append(ret, v) }
    
    return ret
}

File: LeetCode/Medium/0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<int> prime = {2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,79,83,89,97,101};
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> ret;
        unordered_map<long, vector<string>> temp;
        
        for (int i = 0; i < strs.size(); ++i){
            long hash = 1;
            
            for (int j = 0; j < strs[i].size(); ++j) hash = hash * prime[strs[i][j] - 'a'] % 10000000007L;
            
            if (temp.find(hash) == temp.end()) temp[hash] = vector<string>{};
            temp[hash].push_back(strs[i]);
        }
        
        for (auto& it : temp) ret.push_back(it.second);
        
        return ret;
    }
};